lib/galaxy/jobs/__init__.py

```diff
@@ -1721,14 +1706,13 @@ def fail():
         self.tool.call_hook('exec_after_process', self.app, inp_data=inp_data,
                             out_data=out_data, param_dict=param_dict,
                             tool=self.tool, stdout=job.stdout, stderr=job.stderr)
-        job.command_line = unicodify(self.command_line)
 
         collected_bytes = 0
         # Once datasets are collected, set the total dataset size (includes extra files)
         for dataset_assoc in job.output_datasets:
             if not dataset_assoc.dataset.dataset.purged:
-                dataset_assoc.dataset.dataset.set_total_size()
-                collected_bytes += dataset_assoc.dataset.dataset.get_total_size()
+                # don't call get_total_size - forces a flush we don't want in here.
```

Comment:
It's quite confusing that `set_total_size` doesn't save to the database but `get_total_size` does...
Maybe add an explicit `, flush=True` parameter to both?

Reply:
Can I just drop the comment then? It isn't confusing that set_size returns the size right - the confusion is that get size initializes the data but that isn't relevant if the comment weren't here?

Reply:
+1 on @nsoranzo's suggesion. I think, it *is* unexpected that `get_...` may update the database. (just a lazy load with calling the db if value not set would be fine, but to me this feels different).
Maybe not `flush=True`, but `set_if_none=True` or `calculate_if_none`? That way the argument explicitly says *what* will be done (vs. `flush`: what updates will be flushed?)
(BTW, `set_size` does *not* return anything, unless I'm looking in the wrong place).

Reply:
@jmchilton I had nothing against that comment, it was useful. What I have always found confusing is the implementation of both `get_total_size` and `set_total_size` methods. What I am proposing is:

```python
    def get_total_size(self, flush=True):
        if self.total_size is not None:
            return self.total_size
        return self.set_total_size(flush=flush)

    def set_total_size(self, flush=True):
        if self.file_size is None:
            self.set_size()
        self.total_size = self.file_size or 0
        rel_path = self._extra_files_rel_path
        if rel_path is not None:
            if self.object_store.exists(self, extra_dir=rel_path, dir_only=True):
                for root, dirs, files in os.walk(self.extra_files_path):
                    self.total_size += sum([os.path.getsize(os.path.join(root, file)) for file in files if os.path.exists(os.path.join(root, file))])
        if flush:
            db_session = object_session(self)
            db_session.flush()
        return self.total_size
```

That's very probably material for another PR, but what do you think of this?

@ic4f Regarding the parameter name, `flush` is used in all Galaxy performance-sensitive code to mean if a modified object should be flushed to the database or the caller is responsible for that, so I think its meaning would be clear in this context.

Reply:
@nsoranzo thanks for clarifying, this makes sense. +1